contracts/clients/README.md

```diff
@@ -157,34 +155,65 @@ const wallet = await BlsWalletWrapper.connect(
   provider,
 );
 const aggregator = new Aggregator("https://arbitrum-goerli.blswallet.org");
-const estimateFee = await aggregator.estimateFee(bundle); // Remember to include no payment with this bundle
+
+// Create a fee estimate bundle
+const estimateFeeBundle = wallet.sign({
+  nonce,
+  actions: [
+    ...actions, // ... add your user actions here (approve, transfer, etc.)
+    {
+      ethValue: 1,
+      // Provide 1 wei with this action so that the fee transfer to
+      // tx.origin can be included in the gas estimate.
+      contractAddress: aggregatorUtilitiesContract.address,
+      encodedFunction:
+        aggregatorUtilitiesContract.interface.encodeFunctionData(
+          "sendEthToTxOrigin",
+        ),
+    },
+  ],
+});
+
+const feeEstimate = await aggregator.estimateFee(estimateFeeBundle);
+
+// Add a safety premium to the fee to account for fluctuations in gas estimation
+const safetyDivisor = 5;
+const feeRequired = BigNumber.from(feeEstimate.feeRequired);
+const safetyPremium = feeRequired.div(safetyDivisor);
+const safeFee = feeRequired.add(safetyPremium);
 
 const bundle = wallet.sign({
-    nonce: await wallet.Nonce(),
-    actions: [
-        ...actions, // ... do your transaction/actions here (approve, transfer, etc.)
-
-        // then, if the aggregator is using native chain currency, such as ETH
-        {
-            ethValue: estimateFee.feeRequired, // fee amount
-            contractAddress: aggregatorAddress,
-            encodedFunction: "0x"
-        }
-        // or you can use the AggregatorUtilities.sol contract
-        // this makes the bundle submittable from any ETH paid aggregator
-        {
-            ethValue: estimateFee.feeRequired, // fee amount
-            contractAddress: aggregatorUtilitiesContract.address,
-            encodedFunction: aggregatorUtilitiesContract.interface.encodeFunctionData(
-                'sendEthToTxOrigin', [],
-            ),
-        }
-    ],
+  nonce: await wallet.Nonce(),
+  actions: [
+    ...actions, // ... add your user actions here (approve, transfer, etc.)
+    {
+      ethValue: safeFee, // fee amount
+      contractAddress: aggregatorUtilitiesContract.address,
+      encodedFunction:
+        aggregatorUtilitiesContract.interface.encodeFunctionData(
+          "sendEthToTxOrigin",
+        ),
+    },
+
+    // You can also just add an action that transfers the fee amount to the aggregator address, instead of
+    // the action that calls sendEthToTxOrigin. Ensure you modify the estimateFeeBundle to use this action instead.
+    {
+      ethValue: safeFee, // fee amount
+      contractAddress: aggregatorAddress,
+      encodedFunction: "0x",
+    },
+  ],
 });
 ```
 
+```ts
+
+```
+
 ### Paying aggregator fees with custom currency (ERC20)
```

Comment:
@voltrevo one piece of logic that confused me when looking around the code was [this](https://github.com/web3well/bls-wallet/blob/55ad789c3a86831d4bbadff92a402ae1d017f96c/aggregator/test/AggregationStrategy.test.ts#L95). Why is the fee included as the second argument to the mint function? It was my understanding that the second param for mint was the [amount to mint](https://github.com/web3well/bls-wallet/blob/db968c34fd7ce30e01e0a3048aa9075370d367a0/contracts/contracts/mock/MockERC20.sol#L15), not the fee?

I think I've updated the transfer example below to reflect this, but interested to understand the mechanics of how the fee is used with contract interactions

Reply:
Ah yeah, in that test, `mint` is being used to pay the fee. The test doesn't include an action for the user because it's not really relevant to the test - you can think of it like the user is paying for a no-op, which could actually happen if you just wanted to increment your wallet's nonce.

The aggregator can be configured to accept payments in an ERC20 token, which is what's configured for that test ([here](https://github.com/web3well/bls-wallet/blob/55ad789c3a86831d4bbadff92a402ae1d017f96c/aggregator/test/AggregationStrategy.test.ts#L42-L43)). Of course, you'd usually use a transfer to do that, but since we have the authority to mint tokens (because it's a test token), we can simply mint tokens into the aggregator's account.

Reply:
Ahhh gotcha, makes total sense now, thanks for the explanation